Testscripts/Linux/ETHTOOL-OFFLOADING-SETTING.sh

```diff
@@ -0,0 +1,268 @@
+    LogErr "Expected VF count: $NIC_COUNT. Actual VF count: $vf_count"
+    SetTestStateFailed
+    exit 0
+fi
+UpdateSummary "Expected VF count: $NIC_COUNT. Actual VF count: $vf_count"
+
+declare -A offload_features
+offload_features=([tcp-segmentation-offload]=tso [udp-fragmentation-offload]=ufo [generic-segmentation-offload]=gso [generic-receive-offload]=gro [large-receive-offload]=lro [rx-vlan-offload]=rxvlan [tx-vlan-offload]=txvlan [rx-checksumming]=rx [tx-checksumming]=tx)
+
+__iterator=1
+__ip_iterator_1=1
+__ip_iterator_2=2
+
+# Check all the interfaces with VF
+while [ $__iterator -le "$vf_count" ]; do
+    # Extract VF_IP values
+    ip_variable_name="VF_IP$__ip_iterator_1"
+    static_IP_1="${!ip_variable_name}"
+    ip_variable_name="VF_IP$__ip_iterator_2"
+    static_IP_2="${!ip_variable_name}"
+
+    synthetic_interface_vm_1=$(ip addr | grep $static_IP_1 | awk '{print $NF}')
+    LogMsg  "Synthetic interface found: $synthetic_interface_vm_1"
+    if [[ $DISTRO_VERSION =~ ^6\. ]]; then
+        synthetic_MAC=$(ip link show ${synthetic_interface_vm_1} | grep ether | awk '{print $2}')
+        vf_interface_vm_1=$(grep -il ${synthetic_MAC} /sys/class/net/*/address | grep -v $synthetic_interface_vm_1 | sed 's/\// /g' | awk '{print $4}')
+    else
+        vf_interface_vm_1=$(find /sys/devices/* -name "*${synthetic_interface_vm_1}" | grep "pci" | sed 's/\// /g' | awk '{print $12}')
+    fi
+    LogMsg "Virtual function found: $vf_interface_vm_1"
+
+    # Ping the remote host
+    i=0
+    ping_exit_code=1
+    while [ $i -lt 4 ]
+    do
+        i=$(($i+1))
+        sleep 10
+        ping -c 11 "$static_IP_2" > "$static_IP_2.log" 2>&1
+        ping_exit_code=$?
+        LogMsg "Attempt #$i to ping $VF_IP2 through $synthetic_interface_vm_1"
+        if [ "$ping_exit_code" == 0 ]; then
+            LogMsg "Successfully pinged $VF_IP2 through $synthetic_interface_vm_1"
+            break
+        fi
+    done
+
+    if [ "$ping_exit_code" == 1 ]; then
+        LogErr "Unable to ping $VF_IP2 through $synthetic_interface_vm_1"
+        SetTestStateFailed
+        exit 0
+    fi
+
+    # Take a note on the offloading state and set it on
+    for key in ${!offload_features[*]}
+    do
+        on_or_off=$(ethtool -k $synthetic_interface_vm_1 2>&1 | grep $key | awk {'print $2'})
+        fixed_or_not=$(ethtool -k $synthetic_interface_vm_1 2>&1 | grep $key | awk {'print $3'})
+        LogMsg "The initial status of $key for $synthetic_interface_vm_1: $on_or_off $fixed_or_not"
+        # Set the offload on if its status is off and not fixed
+        if [[ "$on_or_off" == "off" && x"$fixed_or_not" != x"[fixed]" ]];then
+            LogMsg "Set ${offload_features[$key]} on"
+            result=$(ethtool -K $synthetic_interface_vm_1 ${offload_features[$key]} on 2>&1)
+            if [[ "$result" = *"Could not change any device features"* ]]; then
+                LogErr "$result"
+                LogErr "The kernel doesn't support this feature"
```

Comment:
should we exit here? I run test against canonical ubuntuserver 18.04-lts Latest, it entered into these lines.

Reply:
Maybe it should be done by 'sudo'. Ubuntu 18.04 should not come this path. If this is the case, then we should ask to @haiyangz

Reply:
The different kernels may support different features. It's not quite right to exit the test if a feature is not supported.

Reply:
> The different kernels may support different features. It's not quite right to exit the test if a feature is not supported.

@xian123 what is your proposal?

Reply:
@haiyangz,

@xian123 @LiliDeng found some of kernel/distro in test scenario do not show the output (It reach to the line 191) Do you think this is the expected case in certain distro, or turn out a bug? Please advise how we handle this failure case.

Reply:
@LiliDeng @Tong-Xu  can you email this question to @haiyangz for the confirmation? We should get @haiyangz feedback for this before close.

Reply:
The pre-condition is seeing NO '[fixed]' string from the output of ethtool -k. Otherwise like before, we would see below.
_tx-scatter-gather: on [fixed]_

So the logic here is checking the string [fixed] in the output first. If it does not show up, @haiyangz patches are applied.